rmw_fastrtps_dynamic_cpp/src/subscription.cpp

```diff
@@ -107,90 +124,149 @@ create_subscription(
+    });
+
   std::string type_name = _create_type_name(
     type_support->data, info->typesupport_identifier_);
-  if (
-    !Domain::getRegisteredType(
-      participant, type_name.c_str(),
-      reinterpret_cast<TopicDataType **>(&info->type_support_)))
-  {
-    info->type_support_ = new (std::nothrow) TypeSupportProxy(type_support_impl);
-    if (!info->type_support_) {
-      RMW_SET_ERROR_MSG("failed to allocate TypeSupportProxy");
-      return nullptr;
-    }
-    _register_type(participant, info->type_support_);
+
+  /////
+  // Register the Type in the participant
+  // When a type is registered in a participant, it is converted to a shared_ptr, so it is
+  // dangerous to keep using it. Thus we use a new TypeSupport created only to register it.
+  ReturnCode_t ret = domainParticipant->register_type(
+    eprosima::fastdds::dds::TypeSupport(new (std::nothrow) TypeSupportProxy(type_support_impl)));
+  // Register could fail if there is already a type with that name in participant,
+  // so not only OK retcode is possible
+  if (ret != ReturnCode_t::RETCODE_OK && ret != ReturnCode_t::RETCODE_PRECONDITION_NOT_MET) {
+    return nullptr;
   }
 
-  if (!participant_info->leave_middleware_default_qos) {
-    subscriberParam.historyMemoryPolicy =
-      eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+  /////
+  // Create Listener
+  info->listener_ = nullptr;
+  if (create_subscription_listener) {
+    info->listener_ = new (std::nothrow) SubListener(info);
   }
 
-  subscriberParam.topic.topicKind = eprosima::fastrtps::rtps::NO_KEY;
-  subscriberParam.topic.topicDataType = type_name;
-  subscriberParam.topic.topicName = _create_topic_name(qos_policies, ros_topic_prefix, topic_name);
+  if (!info->listener_) {
+    RMW_SET_ERROR_MSG("create_subscription() could not create subscription listener");
+    return nullptr;
+  }
+
+  auto cleanup_listener = rcpputils::make_scope_exit(
+    [info]() {
+      delete info->listener_;
+    });
+
+  /////
+  // Create Topic
+
+  // Create Topic name
+  auto topic_name_mangled =
+    _create_topic_name(qos_policies, ros_topic_prefix, topic_name).to_string();
+
+  // Use Topic Qos Default
+  eprosima::fastdds::dds::TopicQos topicQos = domainParticipant->get_default_topic_qos();
 
-  if (!get_datareader_qos(*qos_policies, subscriberParam)) {
+  if (!get_topic_qos(*qos_policies, topicQos)) {
     return nullptr;
   }
 
-  info->listener_ = new (std::nothrow) SubListener(info);
-  if (!info->listener_) {
-    RMW_SET_ERROR_MSG("create_subscriber() could not create subscriber listener");
+  // General function to create or get an already existing topic
+  eprosima::fastdds::dds::TopicDescription * des_topic =
+    rmw_fastrtps_shared_cpp::create_topic_rmw(
+    participant_info,
+    topic_name_mangled,
+    type_name,
+    topicQos);
+
+  if (des_topic == nullptr) {
+    RMW_SET_ERROR_MSG("create_subscription() failed to create topic");
     return nullptr;
   }
 
```

Comment:
There is no `dynamic_cast` to `Topic` Entity just in case there are future specialization of base class `TopicDescription`, as it is done in `publisher.cpp`.

Reply:
As `create_datareader` receives a `TopicDescription*`, the cast is only necessary for writers